daemon/cmd/daemon_main.go

```diff
@@ -1383,29 +1382,6 @@ func initEnv() {
 		log.Fatal("BPF masquerade is not supported for IPv6.")
 	}
 
-	// If there is one device specified, use it to derive better default
-	// allocation prefixes
-	node.InitDefaultPrefix(option.Config.DirectRoutingDevice)
```

Comment:
I see these bits are being removed from `initEnv()` (run before `agentHive.Run()`) and moved later in the initialization order. How late are they moving? How do we know this is safe? I didn't see any discussion about this risk in the commit message.

Reply:
Though this change is not necessary in this PR, I can move them back for the time being.

Reply:
I guess from the hive output, the `InitDefaultPreifx()` call (now `InitLocalNode()` -> `SetDefaultPrefix()`) really just moving after start of pprof, gops, k8s-client, is that how to read this?

```
$ go run ./daemon hive                                                                                                                                                                                             
level=info msg="Memory available for map entries (0.003% of 33378533376B): 83446333B" subsys=config                                                                                                                                                                    
level=info msg="option bpf-ct-global-tcp-max set by dynamic sizing to 292794" subsys=config                                                                                                                                                                            
level=info msg="option bpf-ct-global-any-max set by dynamic sizing to 146397" subsys=config                                                                                                                                                                            
level=info msg="option bpf-nat-global-max set by dynamic sizing to 292794" subsys=config                                                                                                                                                                               
level=info msg="option bpf-neigh-global-max set by dynamic sizing to 292794" subsys=config                                                                                                                                                                             
level=info msg="option bpf-sock-rev-map-max set by dynamic sizing to 146397" subsys=config                                                                                                                                                                             
level=warning msg="Running Cilium with \"kvstore\"=\"\" requires identity allocation via CRDs. Changing identity-allocation-mode to \"crd\"" subsys=config                                                                                                             
Cells:                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                       
  Ⓜ️ agent (Cilium Agent):                                                                                                                                                                                                                                              
      Ⓜ️ infra (Infrastructure):                                                                                                                                                                                                                                        
          Ⓜ️ pprof (pprof HTTP server to expose runtime profiling data):                                                                                                                                                                                                
              🚧 pprof.newServer (cell.go:67):                                                                                                                                                                                                                         
                  ⇨ hive.Lifecycle, logrus.FieldLogger, pprof.Config                                                                                                                                                                                                   
                  ⇦ pprof.Server                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                       
              🛠️ pprof.glob..func1 (cell.go:51): func(pprof.Server)                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                       
          ⚙️ (pprof.Config) {                                                                                                                                                                                                                                           
              Pprof: (bool) false,                                                                                                                                                                                                                                     
              PprofAddress: (string) (len=9) "localhost",                                                                                                                                                                                                              
              PprofPort: (uint16) 6060                                                                                                                                                                                                                                 
          }                                                                                                                                                                                                 
                                                                                                                                                                                                            
                                                                                                                                                                                                            
          Ⓜ️ gops (Gops Agent):                                                                                                                                                                              
              ⚙️ (gops.GopsConfig) {                                                                                                                                                                         
                  GopsPort: (uint16) 9890                                                                                                                                                                   
              }                                                                                                                                                                                             
                                                                                                                                                                                                            
                                                                                                                                                                                                            
              🛠️ gops.registerGopsHooks (cell.go:39): func(hive.Lifecycle, logrus.FieldLogger,                                                                                                               
                gops.GopsConfig)                                                                                                                                                                            
                                                                                                                                                                                                            
          Ⓜ️ k8s-client (Kubernetes Client):                                                                                                                                                                 
              ⚙️ (client.Config) {                                                                                                                                                                           
                  EnableK8s: (bool) true,                                                                                                                                                                   
                  K8sAPIServer: (string) "",                                                                                                                                                                
                  K8sKubeConfigPath: (string) "",                                                     
                  K8sClientQPS: (float32) 0,                                                          
                  K8sClientBurst: (int) 0,                                                            
                  K8sHeartbeatTimeout: (time.Duration) 30s,                                           
                  EnableK8sAPIDiscovery: (bool) false                                                 
              }                                                                                       
                                                                                                      
                                                                                                      
              🚧 client.newClientset (cell.go:112):                                                   
                  ⇨ client.Config, hive.Lifecycle, logrus.FieldLogger                                 
                  ⇦ client.Clientset                                                                  
                                                                                                      
          Ⓜ️ cni-config (CNI configuration manager):                                                   
              ⚙️ (cni.Config) {                                                                        
                  WriteCNIConfWhenReady: (string) "",                                                 
                  ReadCNIConf: (string) "",                                                           
                  CNIChainingMode: (string) (len=4) "none",                                           
                  CNILogFile: (string) (len=30) "/var/run/cilium/cilium-cni.log",                     
                  CNIExclusive: (bool) false                                                          
              }                                                                                       
                                                                                                      
                                                                                                      
              🚧 cni.enableConfigManager (cell.go:58):                                                
                  ⇨ *option.DaemonConfig, cni.Config, hive.Lifecycle, logrus.FieldLogger              
                  ⇦ cni.CNIConfigManager                                                              
                                                                                                      
          🚧 cmd.glob..func1 (cells.go:58):                                                           
              ⇦ *option.DaemonConfig                                                                  
                                                                                                      
      Ⓜ️ controlplane (Control Plane):                                                                 
          Ⓜ️ local-node-store (Provides LocalNodeStore for observing and updating local node info):    
              🚧 node.NewLocalNodeStore (local_node_store.go:69):                                     
                  ⇨ hive.Lifecycle, node.LocalNodeInitializer[optional]                               
                  ⇦ *node.LocalNodeStore                                                              
                                                                                                      
          🚧 cmd.newLocalNodeInitializer (local_node_init.go:53):                                     
              ⇨ *option.DaemonConfig, client.Clientset, k8s.LocalNodeResource                         
              ⇦ node.LocalNodeInitializer

...
```

If I read that right, it moves this to after the k8s apiserver connection is established.  

Then I guess based on the above ordering, the `Start()` function for the local node store is synchronously run (? I didn't see documentation about synchronous assumption in `pkg/hive`), which would then mean that we know that the initialization shouldn't functionally change because it's still ensured to be run before we start handling async events for k8s resources in other Cells. Is that correct?

Reply:
Here's how the start hooks look like:
```
$ go run ./daemon hive --k8s-api-server localhost:1234
..
Start hooks:

  • gops.registerGopsHooks.func1 (cell.go:44)
  • client.(*compositeClientset).onStart
  • authmap.newAuthMap.func1 (cell.go:28)
  • configmap.newMap.func1 (cell.go:24)
  • datapath.newDatapath.func1 (cells.go:85)
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Node].Start
  • node.NewLocalNodeStore.func1 (local_node_store.go:77)
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Service].Start
  • *gobgp.diffStore[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Service].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumNode].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Namespace].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1.CiliumLoadBalancerIPPool].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumIdentity].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumNetworkPolicy].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumClusterwideNetworkPolicy].Start
  • *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1.CiliumCIDRGroup].Start
  • endpointmanager.newDefaultEndpointManager.func1 (cell.go:185)
  • cmd.newPolicyTrifecta.func1 (policy.go:128)
  • *manager.manager.Start
  • monitor.(*dropMonitor).OnStart
  • *cni.cniConfigManager.Start
  • cmd.newDaemonPromise.func1 (daemon_main.go:1599)
  • utime.initUtimeSync.func1 (cell.go:34)
```

We just need to worry about the start hooks before `node.NewLocalNodeStore` accessing anything in pkg/node. This PR drops the default `localNode` instance in pkg/node/address.go, so any uses of the getters will panic so we'll catch those cases. And when `localNode` is set (so that pkg/node getters work again), it's is already partially initialized. And once we've gotten rid of the global getters/setters completely this will be fully driven by the dependency to `LocalNodeStore`.